TEST to look at Gh build vs tip

karianna opened the pull request.

Vest committed:
Added initial GH action support

Vest committed:
Upgraded gradle wrapper to 8.11.1.

Vest committed:
Added testing to gh-action.

Vest committed:
First attempt to log data in test. It failed in GH actions.

Vest committed:
Added the logging to all tests.

Vest committed:
Added gradle cache to a github action setup-java.

Vest committed:
Disabled logging, but added the test results upload.

Vest committed:
Enabled "write" permissions for test results

Vest committed:
Refactored TestHelper.java
Added some logging.

Vest committed:
Updated cache-dependency-path.

Vest committed:
First attempt to cache build/classes and build/jre

Vest committed:
Corrected a typo in gradle.yml

Vest committed:
Try to clear the cache and fill it with values.

Vest committed:
Merge branch 'master' into gh_build

Vest committed:
First step to remove NSIS from the distro;
Removed explicit System.exit(0) in Main. It allows to enable "slowtest" that generates and compares XMLs.
Marked SystemExitInterceptor as deprecated and removal=true. Since Java does not offer any alternative, we have to find a better solution.
P.S. NSIS is planned to be removed to use jpackage only.

Vest committed:
Performance optimization: the debugPrint (when called frequently) uses StackTrace to determine the logger name. This is expensive and replaced with an explicit logger creation.

Vest committed:
Refactored Logging.java.

Vest commented:
Please wait (as usual, I cannot tell until when) and don't merge this PR

Vest committed:
Upgraded gradle-wrapper from 8.11.1 to 8.12

Vest committed:
Remove .run/ from git repository, because it tracked by mistake in #7056

Vest committed:
Refactored several classes. The main goal was adding the Optional to methods that return null. In some cases, the null reference/result was not checked, but the code never failed, because it was not executed.
There were no tests, that is why it was difficult to check, whether the refactoring was right :(

Vest committed:
Merge pull request #2 from PCGen/master

Update gh_build

Vest committed:
Corrected pf_Cleric test;
1. Corrected the typo in special properties (light crossbow);
2. Corrected maxdex for MAXDEX bonus;

Vest committed:
Remove "public" and "abstract" for several interfaces (redundant)

karianna commented:
Yeah I assume we won't merge this PR, I figured it was just the easierst way to look at a diff and see if we could spot anything.

Vest committed:
Merge branch 'master' into gh_build

# Conflicts:
#	data/pathfinder/paizo/roleplaying_game/core_rulebook/cr_equip_arms_armor.lst

karianna commented:
Super interesting those tests fail for you but not for me locally!

karianna commented:
> If you use your version/branch, but not mine, I bet you have this PR: #6975 It comments the most important part, when the character sheet is exported. I have uncommented it, and now the tests started failing. (mostly because LegacyKing corrected some data-files).

OK, so just a data fix required then?

Vest committed:
Added more logging to PcgenFtlTestCase.java (it will help to determine the difference between "golden" and generated files).
Trimmed lines in base.xml and base-xml.ftl template.

Vest committed:
Trimmed different *.xml inputs.
Corrected pf_Rogue.xml result: +1 reflex comes from Halfling Luch and +3 from his Cloak.

Vest commented:
Not only:
1. DATA changes caused slowtests to fail
2. Some XML files were not trimmed
3. The template was trimmed (no leading whitespaces)
4. That's all. I am waiting for the next build :smile:

Vest committed:
Updated build.gradle - disable security manager for all Test Gradle "actions"
Created a "graceful exit" object + interceptor. It is supposed to replace System.Exit calls in different parts of the project;
Updated reporting.gradle by increasing versions of some dependencies;
Removed a DM_EXIT rule from spotbugs -> the calls of System.Exit must be replaced with GracefulExit.
The decision to create a custom interceptor was taken based on the discussion from Reddit: https://www.reddit.com/r/java/comments/1fpxmfp/jep_486_permanently_disable_the_security_manager/

Vest committed:
Added JavaDoc to Exit* and GracefulExit classes.

Vest commented:
Two commits above this comment are related to #7079. Security manager is mostly used to intercept System.exit calls;
I have created required classes to replace System.exit with my proposed class. 

I hope, you will like it.

karianna commented:
Looks like it all passed!  Shall we take out of Draft?

Vest commented:
It is unfinished. Do you insist? I need:
1. remove security manager
2. replace all System.exit calls 
3. I have disabled NSIS, but I have not moved to jpackage a windows build yet. 

If you don’t need release builds for windows for now. I would implement my PR simply to have indicators that tests are green. 

Are you sure? Because I am not. 

You can actually merge it, but I only if you plan to work with the new codebase. E.g., continue improving GH actions. If not, I would keep it for a while draft.

Vest committed:
Resolved warnings for Gradle 8.12: "Space-assignment syntax in Groovy DSL has been deprecated."

Vest committed:
Moved GracefulExit from test to main (util).
Replaced few System.exit with GracefulExit.exit calls.
Refactored few methods to satisfy SonarQube warnings.

Vest committed:
Refactored CommandLineArguments.java to avoid hidden System.exit calls.

Vest committed:
Corrected failed int (slow) tests

Vest committed:
Corrected a wrong assertion message

Vest committed:
Removed SystemExitInterceptor.java, because it was deprecated;
Refactored CommandLineArguments and its test.
Increased the code coverage.

Vest commented:
The commit 59833c5 addresses the issue #7079. We don't use SecurityManager directly (maybe its exceptions are still thrown here and there), but I have created a custom interceptor, called pcgen.util.ExitInterceptor. It is used in several test-classes.
@karianna, if you want, you can link this PR to the mentioned issue #7079 
@Azhrei, I know, you were assigned to this issue. Since I haven't heard any news from you, I think, you were busy with your life. If you are curious about my implementation, can you please review this PR? Not entirely, but you can look at the class: GracefulExit, ExitInterceptor, ExitFunction and few tests that I modified.

I don't know, if I should move my three "exit"-related classes/interfaces into a nested package (just to reduce litter in utils package). If you think, it will be a good idea, I will do this. Something like: "pcgen.util.exit".

Thank you all.

Vest commented:
<img alt="Code coverage of CommandLineArgs" src="https://github.com/user-attachments/assets/1f2eaca3-3109-45d0-be4b-1678ad2e6a80" />

<img alt="Code coverage of GracefulTest" src="https://github.com/user-attachments/assets/4aedf5a0-e620-4364-a168-16af1a3b9fcd" />

Vest committed:
Merge branch 'master' into gh_build

Vest committed:
Added a checkstyle rule that prevents the project from System.exit usage outside GracefulExit.

Vest committed:
Use the latest version of Freemarker template engine.

Azhrei commented:
Thanks.  I actually have some time today, so yes, I'll take a look.  It'll be a couple hours from now, most likely.

Since intercepting System.exit() is a no-no, I'm very curious to see how you did it!

Vest commented:
I don’t intercept it. I have got rid of it. There is only one place, where I allow it - GracefulExit :smile:
The rest is filtered out with CheckStyle. If you accidentally write System.exit, it won’t pass CheckStyle (a simple regexp within all sources - dumb, but solid). You can bypass it, if you overcome my regexp (e.g., use reflection + build the string at runtime). 
thanks.

Vest committed:
Resolved warnings from Gradle.

karianna commented:
@Vest Will happily wait until you're done, thanks for tackling this!

Azhrei commented on code/src/java/pcgen/persistence/lst/LstFileLoader.java:
```diff
@@ -98,40 +96,35 @@ public static String readFromURI(URI uri) throws PersistenceLayerException
 							"The file %s uses UTF-8-BOM encoding. LST files must be UTF-8".formatted(uri));
```
Error message should be in a properties file...
Also, at a higher level, a file that starts with a UTF-8-BOM actually *is* a UTF-8 file, so this whole area might need reworking.

Vest replied:
Nope.
I don't honestly know what to deal with the error here. I have explained you that we had Apache Commons something (maybe String commons, I don't remember exactly) to work with BOM-files. I wanted to optimize the build, and I was trying to get rid of dependencies that are not that important.

When I analyzed all LST files, I found out that we had BOM in (maybe) 10 files, not more. It was easier for me to remove BOM and write code that fails/warns the user, if he gives an unsupported file.

LegacyKing was the first one (we don't have many data-contributors), who stepped this mine first. He was disappointed with the error, and I had to write this code just for him.

My thoughts are to add a Gradle task to go through all LST (or DATA) files and scan for BOM. Either convert them to non-BOM, or fail the build or whatever. 

If I don't throw an error, a wrong LST will be pushed to the repo (I don't want this). If I throw an error, probably this build will fail and somebody else will be unhappy. Maybe I should convert LST files silently, but Karianna didn't like when "gradle test" modified .properties files even when nothing was changed. Maybe I'd better write an integration test that will fail, and after this PR is ready we will be able to see in the pull request, who is trying to break the build. Maybe I will also add a task such as "checkLst" to Gradle, so it will convert broken files to good ones. And in the test exception, I will write something like "BOM detected. Test failed. Use gradle checkLst to repair".

Azhrei replied:
Sounds like you've got it under control. I wasn't aware of the backstory. 👍

Azhrei commented on code/src/java/pcgen/persistence/lst/LstFileLoader.java:
```diff
@@ -98,40 +96,35 @@ public static String readFromURI(URI uri) throws PersistenceLayerException
 				ShowMessageDelegate.showMessageDialog("Preferences are currently set to NOT allow\nloading of "
-					+ "sources from web links.\n" + uri + " is a web link", Constants.APPLICATION_NAME,
-					MessageType.ERROR);
+								+ "sources from web links.\n" + uri + " is a web link", Constants.APPLICATION_NAME,
```
Another properties file template string...?

Vest replied:
Done!

Azhrei reviewed (changes requested):
Wow, big changeset. 🙂  A lot changes not directly related to `GracefulExit`, but I mostly skimmed those.

I like the idea of using GE as a front-end on `System.exit()`.

It all looks good to me!

Vest commented on code/src/java/pcgen/core/PlayerCharacter.java:
```diff
@@ -5477,7 +5477,13 @@ public int processOldMaxDex()
-		int bonus = (load == Load.MEDIUM) ? 2 : (load == Load.HEAVY) ? 1 : (load == Load.OVERLOAD) ? 0 : statBonus;
+		int bonus = switch (load)
+		{
+			case MEDIUM -> 3;
```
Nice catch, but no :) I have found this bug, when my integration tests failed:

<img alt="Encumbrance Effects" src="https://github.com/user-attachments/assets/73c820d4-581a-4493-9a14-f673f7c4e13e" />

Max Dex must be +3

Vest committed:
Resolved comments from Github PR.
Moved a GUI-error message to the localization file.
Formatted error messages

Vest committed:
Merge pull request #3 from Vest/master

Bump com.github.spotbugs from 6.0.28 to 6.1.0 (#7254)

Vest committed:
JaCoCo was added to github actions - now it shows the high-level code coverage my tests

Vest committed:
Merge branch 'master' into gh_build

Vest commented:
@karianna please check my PR: https://github.com/Vest/pcgen/pull/1
Do you like it?

karianna commented:
> @karianna please check my PR: [Vest#1](https://github.com/Vest/pcgen/pull/1) Do you like it?

Looks great!

Vest committed:
Create a jpackage build.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
First attempt to build releases for all OSes.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Changed the event to build a release conditionally.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Update gradle-release.yml

Vest committed:
Corrected the artifact name

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Attempt to create releases

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Attempt to create releases (v2)

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Corrected the artifact creation.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Attempt to create one more release

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest commented:
Ok, I'd say that we have something, that I can call "release": https://github.com/Vest/pcgen/releases/tag/v0.0.8

**Attention:** I am not a professional devops. So, I am experimenting with my workflow. Really. So, it can build releases now. I have tested a mac release only. The application starts. It isn't signed, but it can run.

My thoughts - delete NSIS completely. But I have to verify windows & linux builds somehow.

@karianna

Vest committed:
Updated JDK from 21.0.5 to 21.0.6

Signed-off-by: Vest <Vest@users.noreply.github.com>

karianna commented:
> Ok, I'd say that we have something, that I can call "release": https://github.com/Vest/pcgen/releases/tag/v0.0.8
> 
> **Attention:** I am not a professional devops. So, I am experimenting with my workflow. Really. So, it can build releases now. I have tested a mac release only. The application starts. It isn't signed, but it can run.
> 
> My thoughts - delete NSIS completely. But I have to verify windows & linux builds somehow.
> 
> @karianna

Is it also possible to create .deb linux distro files?  I can test on an Ubuntu system (although it'll be Arm based), I can also test Windows (Arm) but I suspect the exe file is for x86-64?

Azhrei commented:
It is.  GitHub makes a number of environments available for performing the build.  x86_64 is supported for Windows and Linux, and arm64 for Linux and macOS (not sure about Windows).  The Debian (including Ubuntu) and Red Hat (not sure about Fedora) Linux distros are supported.

It's not difficult to create a matrix of supported architectures and supported platforms, then let GitHub build ALL OF THEM!  Obviously, you'd want to be careful about triggering that one!

Vest committed:
Corrected a build for Windows (at least).
gradle jpackageImage

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Added ubuntu-24-arm

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Merge branch 'master' into gh_build

Vest committed:
Added a hotfix to repair linux-arm build.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Merge branch 'refs/heads/master' into gh_build

Vest committed:
Merge branch 'refs/heads/master' into gh_build

Vest committed:
Update Gradle wrapper to version 8.13 and fix output formatting in gradlew

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Merge branch 'refs/heads/master' into gh_build

Vest committed:
Update Gradle wrapper to version 8.14

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Removed jar-export-plugins from pluginbuild.xml
Added the same task to plugins.gradle

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Update Appveyor config to adjust Gradle commands and formatting

Removed the `buildNsis` task from the Gradle `test_script`, streamlining the build process. Additionally, fixed unnecessary trailing spaces in multiple sections for better formatting consistency. These changes simplify the CI process and improve readability.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Remove Ant-based build files for Gradle-only build system

This commit deletes `build-gradle.xml` and `pluginbuild.xml` as they are no longer required. The transition to a Gradle-based build system eliminates dependency on Ant, simplifying the build process.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Update properties and refine Gradle plugin Jar tasks

Added a new warning message for remote list preferences in properties. Improved Gradle tasks to include all nested class files for export and bonus plugins, ensuring more comprehensive builds.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Refactor code style by removing unnecessary whitespace and updating method signatures in PCGenTaskExecutor and PluginLoader

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Refactored and changed files that didn't pass ```gradle checkstyleMain```
Few recent pull requests broke the styling: https://github.com/PCGen/pcgen/pull/7319

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Refactored and changed files that didn't pass ```gradle checkstyleMain```
Few recent pull requests broke the styling: https://github.com/PCGen/pcgen/pull/7319

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Refactor PluginClassLoader

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Add plugins.gradle to Gradle configuration for release and test workflows

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Merge branch 'master' into gh_build

# Conflicts:
#	build.gradle

Vest committed:
Update Gradle wrapper to version 9.0.0 and fix classpath configuration in wrapper scripts

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Update JavaFX version to 21.0.8 and improve URIFactory error handling with new tests

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Improve README formatting and grammar for clarity and consistency

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Refactor test task configurations to use correct directory paths for test classes

Signed-off-by: Vest <Vest@users.noreply.github.com>

LegacyKing commented:
Where do we stand with this?

Vest committed:
Changed PluginBuildTest, because the new plugins are built from corresponding .java files.
The slow test has been reverted, because it was broken in old commits (DATA-4372)

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Merge branch 'master' into gh_build

# Conflicts:
#	build.gradle

Vest committed:
Updated Java to 21.0.8 (mac aarch64)
Gradle wrapper to 9.1.0

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Merge branch 'master' into gh_build

# Conflicts:
#	code/src/java/pcgen/core/PlayerCharacter.java
#	code/src/java/pcgen/io/ExportHandler.java

Vest committed:
Update Gradle wrapper to version 9.2.0 and refactor dependency declarations to avoid Gradle's warnings. Such constructions will be deprecated in Gradle 10.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Update Java&JavaFX to 25.0.1 and JaCoCo to 0.8.14 versions in build.gradle

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Add native access flags for JavaFX modules in distribution.gradle

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Refactor JavaFX resource loading by introducing PanelFromResource utility class

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Reverted a part of #7462, that broke DescriptionTest.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Update FXML files to use JavaFX 25.0.1 namespace

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Merge branch 'master' into gh_build

Vest committed:
Corrected checkstyle findings

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Updated PMD to 7.17
Added many rules to ruleset.xml
Fixed few warnings
Enabled config-cache for testing purposes.

Signed-off-by: Vest <Vest@users.noreply.github.com>

Vest committed:
Merge branch 'master' into gh_build

# Conflicts:
#	build.gradle
